### PR TITLE
enable config overrides in hf model

### PR DESCRIPTION
**Pull request by vchiley**
This PR
- enable config overrides in hf model
- removes `device: cpu  # this is where to initialize weights; cpu memory or meta tensor` from HF configs (since HF models do not actually enable us to specify device)

### CODE REVIEW

**Comment by abhi-mosaic** [comment]
I think `device: cpu` will come in handy in the future, we want to be able to load giant HF models and we have some plans to add a `with meta_init` context manager. But I'm ok with adding it back when its ready...

File: examples/llm/yamls/hf_causal_lm/gpt-neo-125m.yaml [hunk 1]
```diff
@@ -11,7 +11,10 @@ run_name: gpt-neo-125m
 model:
   name: hf_causal_lm
   pretrained_model_name_or_path: EleutherAI/gpt-neo-125M
-  device: cpu  # this is where to initialize weights; cpu memory or meta tensor
+  config_overrides:
+    # WARNING: if setting `pretrained: true`, `max_position_embeddings` must match the 
+    # `max_position_embeddings` used durring pre-training
```
Comment (dakinggg) [inline examples/llm/yamls/hf_causal_lm/gpt-neo-125m.yaml:16]:
```suggestion
    # `max_position_embeddings` used during pre-training
```

File: examples/llm/yamls/hf_causal_lm/gpt2-small.yaml [hunk 2]
```diff
@@ -11,7 +11,10 @@ run_name: gpt2-small
 model:
   name: hf_causal_lm
   pretrained_model_name_or_path: gpt2
-  device: cpu  # this is where to initialize weights; cpu memory or meta tensor
+  config_overrides:
+    # WARNING: if setting `pretrained: true`, `max_position_embeddings` must match the 
+    # `max_position_embeddings` used durring pre-training
```
Comment (dakinggg) [inline examples/llm/yamls/hf_causal_lm/gpt2-small.yaml:16]:
```suggestion
    # `max_position_embeddings` used during pre-training
```

File: examples/llm/yamls/hf_causal_lm/gpt2-small.yaml [hunk 3]
```diff
@@ -72,10 +75,11 @@ algorithms:
 
 max_duration: 4800ba # ~ 2.5B tokens
 eval_interval: 500ba
++eval_subset_num_batches: -1
```
Comment (dakinggg) [inline examples/llm/yamls/hf_causal_lm/gpt2-small.yaml:78]:
```suggestion
eval_subset_num_batches: -1
```

Comment (vchiley) [reply examples/llm/yamls/hf_causal_lm/gpt2-small.yaml]: ty

File: examples/llm/yamls/hf_causal_lm/gpt-neo-125m.yaml (hunk 1 above)
Comment (vchiley) [reply examples/llm/yamls/hf_causal_lm/gpt-neo-125m.yaml]: ty

File: examples/llm/yamls/hf_causal_lm/gpt2-small.yaml (hunk 2 above)
Comment (vchiley) [reply examples/llm/yamls/hf_causal_lm/gpt2-small.yaml]: ty

File: examples/llm/yamls/hf_causal_lm/gpt-neo-125m.yaml [hunk 4]
```diff
@@ -73,10 +75,11 @@ algorithms:
 
 max_duration: 4800ba # ~ 2.5B tokens
 eval_interval: 500ba
+eval_subset_num_batches: -1
 global_train_batch_size: 256
 
 # System
-seed: 17
+seed: *global_seed
```
Comment (dakinggg) [inline examples/llm/yamls/hf_causal_lm/gpt-neo-125m.yaml:82]: make sure you or abhi syncs this up with the yaml changes in his pr

File: examples/llm/yamls/hf_causal_lm/gpt2-small.yaml [hunk 5]
```diff
@@ -72,10 +75,11 @@ algorithms:
 
 max_duration: 4800ba # ~ 2.5B tokens
 eval_interval: 500ba
+eval_subset_num_batches: -1
 global_train_batch_size: 512
 
 # System
-seed: 17
+seed: *global_seed
```
Comment (dakinggg) [inline examples/llm/yamls/hf_causal_lm/gpt2-small.yaml:82]: same

File: examples/llm/yamls/hf_causal_lm/gpt-neo-125m.yaml [hunk 6]
```diff
@@ -11,7 +11,10 @@ run_name: gpt-neo-125m
 model:
   name: hf_causal_lm
   pretrained_model_name_or_path: EleutherAI/gpt-neo-125M
-  device: cpu  # this is where to initialize weights; cpu memory or meta tensor
+  config_overrides:
+    # WARNING: if setting `pretrained: true`, `max_position_embeddings` must match the 
+    # `max_position_embeddings` used during pre-training
+    max_position_embeddings: *max_seq_len
```
Comment (dakinggg) [inline examples/llm/yamls/hf_causal_lm/gpt-neo-125m.yaml:17]: same

File: examples/llm/yamls/hf_causal_lm/gpt2-small.yaml [hunk 7]
```diff
@@ -11,7 +11,10 @@ run_name: gpt2-small
 model:
   name: hf_causal_lm
   pretrained_model_name_or_path: gpt2
-  device: cpu  # this is where to initialize weights; cpu memory or meta tensor
+  config_overrides:
+    # WARNING: if setting `pretrained: true`, `max_position_embeddings` must match the 
+    # `max_position_embeddings` used during pre-training
+    n_positions: *max_seq_len
```
Comment (dakinggg) [inline examples/llm/yamls/hf_causal_lm/gpt2-small.yaml:17]: same

**Comment by abhi-mosaic** [comment]
Probably worth rebasing this on `main` now, @vchiley

**Review by abhi-mosaic — APPROVED** [review]
This looks good to me, but would you like to change the `gpt2-small` YAML to use a `max_seq_len: 2048` by default to match all the other models?  Because this is made possible by your fix :)

**Pull request merged**